libraries/ZigBeeNet/Database/ZigBeeNetworkDatabaseManager.cs

```diff
@@ -0,0 +1,292 @@
+                return;
+            }
+            _deferredWriteTime = deferredWriteTime;
+        }
+
+        /// <summary>
+        /// Sets the maximum deferred write timer. This will ensure that the data store is updated periodically in the event
+        /// that continuous updates are received for a node.
+        /// </summary>
+        /// <param name="deferredWriteMax">the maximum number of milliseconds that writes will be deferred</param>
+        public void SetMaxDeferredWriteTime(int deferredWriteMax) 
+        {
+            if (deferredWriteMax > DEFERRED_WRITE_TIMEOUT_MAX) 
+            {
+                _deferredWriteTimeout = DEFERRED_WRITE_TIMEOUT_MAX;
+                return;
+            }
+            _deferredWriteTimeout = deferredWriteMax;
+        }
+
+        /// <summary>
+        /// Clears all data from the data store. This may be used when initialising a network to remove any previous data.
+        /// </summary>
+        public void Clear() 
+        {
+            if (DataStore == null) 
+            {
+                Log.Debug("Data store: Undefined so network is not cleared.");
+                return;
+            }
+
+            Log.Debug("Data store:  Clearing all nodes.");
+            ISet<IeeeAddress> nodes = DataStore.ReadNetworkNodes();
+            foreach (IeeeAddress nodeAddress in nodes) 
+            {
+                DataStore.RemoveNode(nodeAddress);
+            }
+        }
+
+        /// <summary>
+        /// Starts the database manager. This will call the ZigBeeNetworkDataStore to retrieve the list of nodes, and
+        /// then read all the nodes from the store, adding them to the ZigBeeNetworkManager.
+        /// </summary>
+        public void Startup() 
+        {
+            if (DataStore == null) 
+            {
+                Log.Debug("Data store: Undefined so network is not restored.");
+                return;
+            }
+
+            ISet<IeeeAddress> nodes = DataStore.ReadNetworkNodes();
+            foreach (IeeeAddress nodeAddress in nodes) 
+            {
+                ZigBeeNode node = new ZigBeeNode(_networkManager, nodeAddress);
+                ZigBeeNodeDao nodeDao = DataStore.ReadNode(nodeAddress);
+                if (nodeDao == null) 
+                {
+                    Log.Debug("{IeeeAddress}: Data store: Node was not found in database.", nodeAddress);
+                    continue;
+                }
+                node.SetDao(nodeDao);
+                Log.Debug("{IeeeAddress}: Data store: Node was restored.", nodeAddress);
+                _networkManager.UpdateNode(node);
+            }
+
+            _nodesToSave = new BlockingCollection<ZigBeeNode>();
+
+            //The consumer thread that process the nodes produced by the elapsed timers
+            //This ensure all write are done using a single thread.
+            Thread thread = new Thread(new ThreadStart(WriteNodeLoop));
```

Comment:
Better to use a Task instead of thread, because of internal logic
https://stackoverflow.com/questions/13429129/task-vs-thread-differences

Reply:
done